r/src/arrow_cpp11.h

```diff
@@ -148,7 +148,7 @@ inline SEXP utf8_strings(SEXP x) {
 
     for (R_xlen_t i = 0; i < n; i++, ++p_x) {
       SEXP s = *p_x;
-      if (s != NA_STRING) {
+      if (s != NA_STRING && ALTREP(s)) {
```

Comment:
This does not actually work yet. With this https://github.com/apache/arrow/blob/c3ebdf500e75ca868f50b7d374fc8ce2237756b8/r/tests/testthat/test-utf.R#L19-L38 fail.

Before #43173 this line was:

```
if (s != NA_STRING && !IS_UTF8(s) && !IS_ASCII(s)) {
```

I suspect what's going on is that we caught vroom altrep vectors with this if and called `SET_STRING_ELT` so we didn't get the `No Set_elt found for ALTSTRING` error. But which `ALTREP(s)` we detect the non-utf strings here too and attempt to `SET_STRING_ELT` when we shouldn't.

Reply:
Any thoughts @nealrichardson ?

Reply:
I guess I don't understand what's happening in here. L144 says that this should no longer be altrep by the time we get here.

Reply:
I also think that `STRING_PTR_RO()` should materialize an ALTREP character vector, otherwise you can't iterate over the `CHARSXP` pointers. 

Also, `ALTREP(s)` does not seem to make sense to me, that's a `CHARSXP` (not a `STRSXP`), so it cannot be ALTREP, unless I am misremembering something.

Are you trying to catch the non-UTF-8 strings here? `Rf_translateCharUTF8()` does not do anything (returns the same `const char *`) if the string is UTF-8, so you could always call it, and only call `SET_STRING_ELT()` if the returned pointer is different?

Reply:
> I also think that STRING_PTR_RO() should materialize an ALTREP character vector, otherwise you can't iterate over the CHARSXP pointers.

Hmm, so maybe the actual problem here is that that isn't working on the ALTREP vectors coming from vroom? Here is the error we started seeing when we changed this condition from `s != NA_STRING && !IS_UTF8(s) && !IS_ASCII(s)` to `s != NA_STRING`:

```
Error in Table__from_dots(dots, schema, option_use_threads()): No Set_elt found for ALTSTRING class [class: vroom_chr, pkg: vroom]
```

> Also, ALTREP(s) does not seem to make sense to me, that's a CHARSXP (not a STRSXP), so it cannot be ALTREP, unless I am misremembering something.

Yeah, I agree that `ALTREP(s)` here is not right. 

> Are you trying to catch the non-UTF-8 strings here? Rf_translateCharUTF8() does not do anything (returns the same const char *) if the string is UTF-8, so you could always call it, 

I will admit I'm not 100% certain what is going on here, but I catching non-UTF-8 strings here is what I thought it was doing. 

> and only call SET_STRING_ELT() if the returned pointer is different?

Oh interesting. Forgive my C++ naïveness, but would this be something like (i.e. does `!=` check if the pointer is different>):

```
SEXP new_s = Rf_mkCharCE(Rf_translateCharUTF8(s);
if (new_s != s) {
    SET_STRING_ELT(x, i, Rf_mkCharCE(Rf_translateCharUTF8(new_s), CE_UTF8));
}
```

Reply:
Yeah, that C++ looks correct. 

However, I think the issue is that if `x` is ALTREP, then it does not matter if it is materialized of not, it is still ALTREP, so you still cannot change it with `SET_STRING_ELT`. 

To avoid this, you'd need to call `Rf_duplicate()` on `x`. That should create a non-altrep copy of it. 

You can always call `Rf_duplicate()`, or you can call it when you encounter a non-utf8 string. Note that you need to `PROTECT()` the result of `Rf_duplicate()`

Reply:
Aaah thanks for those pointers! That looked like it worked, I'm going to trigger a benchmark run to see if there are any unintended consequences there (Hopefully no, but even if so, we might need to accept them!)